File: leet/0077-combinations/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        if n == 0 or k == 0 : return []
        
        ret = []
        stack = []
        
        def dfs(nn: int, kk: int):
            if kk == 0 : 
                ret.append([] + stack)
                return
        
        
            for i in range(nn, 0, -1):
                stack.append(i)
                dfs(i-1, kk-1)
                stack.pop()
                
        dfs(n, k)
        return ret        

File: leet/0077-combinations/combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    if(n == 0 || k == 0) return []

    let nums = new Array(n)
    for (let i = 0; i < n; i++) nums[i] = i + 1

    let stack = []
    let ret = []

    let bt = function(i){
        if (stack.length == k){
            ret.push(stack.concat())
            return
        }

        for (;i < n ; i++){
            stack.push(nums[i])
            bt(i + 1)
            stack.pop()
        }
    }

    bt(0)
    return ret
};
